### PR TITLE
Suggestions for the tests

### PR DESCRIPTION
**Pull request by akauppi**
Same game as last time. I'll make some comments and ping you when the PR's ready for review.

It's not as clean as I was hoping. Further PR's will be better. But have a look, check if there's some ideas you like. Don't pass the comments into the code.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -20,17 +18,14 @@ scalacOptions ++= Seq(
...
   "com.typesafe.akka" %% "akka-actor" % "2.3.11",
-  "com.google.guava" % "guava" % "19.0",
+  "com.typesafe.akka" %% "akka-http-experimental" % "2.0.2",
   "com.typesafe.scala-logging" %% "scala-logging" % "3.1.0",
+  "ch.qos.logback" % "logback-classic" % "1.1.3",
   "com.fasterxml.jackson.core" % "jackson-core" % "2.7.0",
   "com.fasterxml.jackson.module" %% "jackson-module-scala" % "2.6.3",
-  "ch.qos.logback" % "logback-classic" % "1.1.3",
-  "com.typesafe.akka" %% "akka-http-experimental" % "2.0.2",
+  "com.google.guava" % "guava" % "19.0",
```
Comment (akauppi) [inline build.sbt]: (minor) I reordered the libraries, to group akka, logging, together.

File: src/test/scala/org/zalando/nakadi/client/KlientBuilderSpec.scala [hunk 2]
```diff
@@ -8,21 +8,33 @@ import org.scalatest.{WordSpec, Matchers}
 class KlientBuilderSpec extends WordSpec with Matchers {
 
   "A Klient builder" must {
-    "must build a Klient instance, if everything is set properly" in {
+
```
Comment (akauppi) [inline src/test/scala/org/zalando/nakadi/client/KlientBuilderSpec.scala]: The "must" need not be repeated in the individual test messages. Note: it is really important what test messages say, because they may be read by not only programmers (this is one of the aspects we're trying to improve in Helsinki - anyone should be able to understand, what was tested, what failed etc.)

File: src/test/scala/org/zalando/nakadi/client/KlientBuilderSpec.scala [hunk 3]
```diff
@@ -8,21 +8,33 @@ import org.scalatest.{WordSpec, Matchers}
...
+    // These things get tested by other things, anyways, so I'm not sure it's useful to test it (in development phase,
+    // it's probably been useful). AKa280116
+    //
+    "build a Klient instance, if everything is set up properly" in {
```
Comment (akauppi) [inline src/test/scala/org/zalando/nakadi/client/KlientBuilderSpec.scala]: The "must" need not be repeated in the individual test messages. 
Note (slightly unrelated): it is really important what test messages say, because they may be read by not only programmers (this is one of the aspects we're trying to improve in Helsinki - anyone should be able to understand, what was tested, what failed etc.)

File: src/test/scala/org/zalando/nakadi/client/KlientSpec.scala [hunk 4]
```diff
@@ -9,53 +9,24 @@ import com.fasterxml.jackson.module.scala.DefaultScalaModule
...
-    while(! receivedEvents.compareAndSet(old, old ++ List(event)))
-  }
-  override def onConnectionClosed(topic: String, partition: String, lastCursor: Option[Cursor]): Unit = onConnectionClosed += 1
-  override def onConnectionOpened(topic: String, partition: String): Unit = onConnectionOpened += 1
-  override def onConnectionFailed(topic: String, partition: String, status: Int, error: String): Unit = onConnectionFailed += 1
-}
-
-
-class KlientSpec extends WordSpec with Matchers with BeforeAndAfterEach with LazyLogging {
+class KlientSpec extends WordSpec with Matchers with BeforeAndAfterEach with LazyLogging with ScalaFutures {
```
Comment (akauppi) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: oops, the ScalaFutures sneaked in here. The code's not using it, yet. It provides an alternative way to do what `Await.result` does. Will make a separate PR on that.

Comment (akauppi) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: Oops. The ScalaFutures is for another PR coming up. It does not hurt, but this code does not need it either.

File: src/test/scala/org/zalando/nakadi/client/KlientSpec.scala [hunk 5]
```diff
@@ -9,53 +9,24 @@ import com.fasterxml.jackson.module.scala.DefaultScalaModule
...
 import com.typesafe.scalalogging.LazyLogging
 import io.undertow.util.{HeaderValues, HttpString, Headers}
-import org.scalatest.{BeforeAndAfterEach, Matchers, WordSpec}
+import org.scalatest._
+import org.scalatest.concurrent.PatienceConfiguration.{Timeout => ScalaTestTimeout}
+import org.scalatest.concurrent.ScalaFutures
 import org.zalando.nakadi.client.actor.PartitionReceiver
 import org.zalando.nakadi.client.utils.NakadiTestService
 import org.zalando.nakadi.client.utils.NakadiTestService.Builder
+import org.zalando.nakadi.client.utils.{Request => NakadiRequest}
```
Comment (akauppi) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: I prefer a bit longer aliases to simply `Request`, when the domain does have other possible alternatives for it (like HTTP Request).

File: src/test/scala/org/zalando/nakadi/client/KlientSpec.scala [hunk 6]
```diff
@@ -9,53 +9,24 @@ import com.fasterxml.jackson.module.scala.DefaultScalaModule
 import com.google.common.collect.Iterators
 import com.typesafe.scalalogging.LazyLogging
 import io.undertow.util.{HeaderValues, HttpString, Headers}
-import org.scalatest.{BeforeAndAfterEach, Matchers, WordSpec}
+import org.scalatest._
+import org.scalatest.concurrent.PatienceConfiguration.{Timeout => ScalaTestTimeout}
```
Comment (akauppi) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: oops. To another PR. ignore :)

File: src/test/scala/org/zalando/nakadi/client/KlientSpec.scala [hunk 7]
```diff
@@ -74,44 +46,29 @@ class KlientSpec extends WordSpec with Matchers with BeforeAndAfterEach with Laz
...
+  private def performStandardRequestChecks(expectedRequestPath: String, expectedRequestMethod: HttpString): NakadiRequest = {
     val collectedRequestsMap = service.getCollectedRequests
     val requests = collectedRequestsMap.get(expectedRequestPath)
     requests should not be null
 
     val request = Iterators.getLast(requests.iterator)
-    request.getRequestPath should be(expectedRequestPath)
-    request.getRequestMethod should be(expectedRequestMethod)
+      //
+      request.getRequestPath shouldBe expectedRequestPath
```
Comment (akauppi) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: I slightly prefer `shouldBe X` over `should be (x)` since it does not require the paranthesis.

File: src/test/scala/org/zalando/nakadi/client/KlientSpec.scala [hunk 8]
```diff
@@ -74,44 +46,29 @@ class KlientSpec extends WordSpec with Matchers with BeforeAndAfterEach with Laz
...
+      request.getRequestMethod shouldBe expectedRequestMethod
 
     val headerMap = request.getRequestHeaders
 
-    var headerValues: HeaderValues = null
```
Comment (akauppi) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: There was really no need for the `var`. I may have shortened it slightly too much.

File: src/test/scala/org/zalando/nakadi/client/KlientSpec.scala [hunk 9]
```diff
@@ -74,44 +46,29 @@ class KlientSpec extends WordSpec with Matchers with BeforeAndAfterEach with Laz
...
-
-    headerValues = headerMap.get(Headers.AUTHORIZATION)
-    val authorizationHeaderValue = headerValues.getFirst
-    authorizationHeaderValue should be(s"Bearer $TOKEN")
+    headerMap.get(Headers.AUTHORIZATION).getFirst shouldBe s"Bearer $TOKEN"
 
     request
   }
 
-  private def checkQueryParameter(queryParameters: java.util.Map[String, util.Deque[String]], paramaterName: String, expectedValue: String) {
```
Comment (akauppi) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: moved this closer to where it's used (only used in one place)

File: src/test/scala/org/zalando/nakadi/client/KlientSpec.scala [hunk 10]
```diff
@@ -353,11 +310,11 @@ class KlientSpec extends WordSpec with Matchers with BeforeAndAfterEach with Laz
       receivedEvents should contain(event2)
 
       listener.onConnectionOpened should be > 0
-      listener.onConnectionClosed should be > 0 // no long polling actitvated by test mock
+      listener.onConnectionClosed should be > 0 // no long polling activated by test mock
 
 
       val collectedRequests = service.getCollectedRequests
-      collectedRequests should have size 3
+      collectedRequests.size shouldBe 3
```
Comment (akauppi) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: What you have is quite fine. However, I tend to prefer a hands-on subset of what test frameworks offer. This makes the learning curve for programmers slightly smaller. Official documents may suggest `should have size N` (because it's there) but I think it's a slightly unneeded feature. What does `have` mean, anyways? (i.e. one needs to go read the docs :) )

File: src/test/scala/org/zalando/nakadi/client/KlientSpec.scala [hunk 11]
```diff
@@ -515,3 +485,53 @@ class KlientSpec extends WordSpec with Matchers with BeforeAndAfterEach with Laz
     service.start()
   }
 }
+
+object KlientSpec extends WordSpecLike /*for 'info()'*/ with ShouldMatchers {
```
Comment (akauppi) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: One can pull aside testing tools to a companion object (so people can start reading the tests, early on in the file). By deriving from `WordSpecLike` we get `info` for providing additional data about particular tests (println-like, but slightly different - see below).

File: src/test/scala/org/zalando/nakadi/client/KlientSpec.scala [hunk 12]
```diff
@@ -74,44 +44,28 @@ class KlientSpec extends WordSpec with Matchers with BeforeAndAfterEach with Laz
     }
   }
 
-  private def performStandardRequestChecks(expectedRequestPath: String, expectedRequestMethod: HttpString) = {
+  private def performStandardRequestChecks(expectedRequestPath: String, expectedRequestMethod: HttpString): NakadiRequest = {
```
Comment (akauppi) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: Providing the return value instead of relying on type inference is important for two reasons:
- to inform the human reader on what to expect, without needing to read the body
- to encapsulate compile errors e.g. deriving of refactoring. Without giving the return type, changing the function will cause changes to ripple to its callers, messing up their syntax highlighting (and more serious stuff).

File: src/test/scala/org/zalando/nakadi/client/KlientSpec.scala (hunk 11 above)
Comment (codechimp) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: nice :-)

File: src/test/scala/org/zalando/nakadi/client/KlientSpec.scala [hunk 13]
```diff
@@ -515,3 +484,53 @@ class KlientSpec extends WordSpec with Matchers with BeforeAndAfterEach with Laz
...
+
+    override def onReceive(topic: String, partition: String, cursor: Cursor, event: Event): Unit = {
+
+      // Note: Changing 'println' to 'info' is a functional change. What is the intended action here?
+      //    - for giving information about tests being run, 'info' (or warning, etc.) are right. Their output
+      //      is placed alongside the test pass/fail reports, in the right location. 'println', on the other hand,
+      //      causes output not aligned with test reporting. IF things are intended to be visible "right then",
+      //      using 'System.err.println' could be yet another choice. AKa280116
+
+      info(s"WAS CALLED [topic=$topic, partition=$partition, event=$event]")
```
Comment (codechimp) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: where does info come from? actually, it has no real meaning. I added this output while I was debugging an error and I forgot to remove it...

File: src/test/scala/org/zalando/nakadi/client/KlientSpec.scala (hunk 12 above)
Comment (codechimp) [inline src/test/scala/org/zalando/nakadi/client/KlientSpec.scala]: I agree. I usually do this with public functions. Sometimes , I do it for private functions as well..depends on the context but I am fine with always specifying it

**Comment by codechimp** [comment]
:+1:  I really like your ideas. From my side, we can merge it
